src/Adaptivity/AdaptedDiscreteModels.jl

```diff
@@ -0,0 +1,282 @@
+
+"""
+  AdaptedDiscreteModel
+
+  `DiscreteModel` created by refining/coarsening another `DiscreteModel`. 
+  The refinement/coarsening hierarchy can be traced backwards by following the 
+  `parent` pointer chain. This allows the transfer of dofs 
+  between `FESpaces` defined on this model and its ancestors.
+
+"""
+struct AdaptedDiscreteModel{Dc,Dp,A<:DiscreteModel{Dc,Dp},B<:DiscreteModel{Dc,Dp},C<:AdaptivityGlue} <: DiscreteModel{Dc,Dp}
+  model  ::A
+  parent ::B
+  glue   ::C
+
+  function AdaptedDiscreteModel(model::DiscreteModel{Dc,Dp},parent,glue) where {Dc,Dp}
+    @check !isa(model,AdaptedDiscreteModel)
+    A = typeof(model)
+    B = typeof(parent)
+    C = typeof(glue)
+    return new{Dc,Dp,A,B,C}(model,parent,glue)
+  end
+end
+
+# DiscreteModel API
+Geometry.get_grid(model::AdaptedDiscreteModel)          = get_grid(model.model)
+Geometry.get_grid_topology(model::AdaptedDiscreteModel) = get_grid_topology(model.model)
+Geometry.get_face_labeling(model::AdaptedDiscreteModel) = get_face_labeling(model.model)
+
+# Other getters
+get_model(model::AdaptedDiscreteModel)  = model.model
+get_parent(model::AdaptedDiscreteModel{Dc,Dp,A,<:AdaptedDiscreteModel}) where {Dc,Dp,A} = get_model(model.parent)
+get_parent(model::AdaptedDiscreteModel{Dc,Dp,A,B}) where {Dc,Dp,A,B} = model.parent
+get_adaptivity_glue(model::AdaptedDiscreteModel) = model.glue
+
+# Relationships
+"""
+Returns true if m1 is a "child" model of m2, i.e., if m1 is the result of adapting m2
+"""
+function is_child(m1::AdaptedDiscreteModel,m2::DiscreteModel)
+  return get_parent(m1) === m2 # m1 = refine(m2)
+end
+
+function is_child(m1::AdaptedDiscreteModel,m2::AdaptedDiscreteModel)
+  return get_parent(m1) === get_model(m2) # m1 = refine(m2)
+end
+
+is_child(m1::DiscreteModel,m2::AdaptedDiscreteModel) = false
+
+is_related(m1::DiscreteModel,m2::DiscreteModel) = is_child(m1,m2) || is_parent(m1,m2)
+
+# Model Refining
+function refine(model::DiscreteModel,args...;kwargs...) :: AdaptedDiscreteModel
+  @abstractmethod
+end
+
+function refine(model::AdaptedDiscreteModel,args...;kwargs...)
+  ref_model = refine(model.model,args...;kwargs...)
+  return AdaptedDiscreteModel(ref_model.model,model,ref_model.glue)
+end
+
+
+# UnstructuredDiscreteModel Refining
+
+function refine(model::UnstructuredDiscreteModel{Dc,Dp}) where {Dc,Dp}
```

Comment:
A question that comes into my mind here. To-think. 

Would it make sense that the `refine` method gets a `RefinementRule` instance?

At first glance I am concerned the `RefinementRule` is not there, but perhaps I am missing something.

Reply:
Yes. This routines predate `RefinementRule`. I maybe have to give some thought on how to re-make them. They are also half-implemented, only focusing on very simple 2D cases. Maybe I could also expand them so that they are a little more complete.

Reply:
FYI ... you can, e.g., see a formal definition of a particular refinement rule here:

https://bonndoc.ulb.uni-bonn.de/xmlui/handle/20.500.11811/7661

Section 4.1.1.

I am not sure we want to introduce all this complexity at this stage, but just to keep in mind that a refinement rule can be seen as an abstraction that prescribes how we subdivide the reference polytope into subpolytopes + the geometric/topologic relationships among the coarser and finer d-faces (vertices, edges, faces, etc.)

At present, I think this abstraction is not explicit, we are just implementing a particular refinement rule from the set of possible refinement rules. 

Do you understand my point? I think there is not action to be taken here at this stage, just to keep in mind ...

Reply:
>Maybe I could also expand them so that they are a little more complete. 

I would live as this as future work (+ an issue describing these discussions). At the end of the day, you need this `refine` method just to able to write mock tests of all the machinery you have written.

Reply:
Another concern that I have here in regards to the refinement of meshes with simplices (triangles, tetrahedra). 

In Gridap, the FE machinery, assumes (requires) that simplices meshes are properly oriented. This is achieved by sorting the global vertex identifiers within each cell in increasing order.

See, e.g., https://github.com/gridap/GridapGmsh.jl/blob/master/src/GmshDiscreteModels.jl#L311

Are we fulfilling this condition for the meshes resulting from refinement? (I did not check it)

See also section 4.3 in the following manuscript https://www.sciencedirect.com/science/article/pii/S096599781831113X

Reply:
> FYI ... you can, e.g., see a formal definition of a particular refinement rule here:
> https://bonndoc.ulb.uni-bonn.de/xmlui/handle/20.500.11811/7661
> Section 4.1.1.
> I am not sure we want to introduce all this complexity at this stage, but just to keep in mind that a refinement rule can be seen as an abstraction that prescribes how we subdivide the reference polytope into subpolytopes + the geometric/topologic relationships among the coarser and finer d-faces (vertices, edges, faces, etc.)
> At present, I think this abstraction is not explicit, we are just implementing a particular refinement rule from the set of possible refinement rules.
> Do you understand my point? I think there is not action to be taken here at this stage, just to keep in mind ...

I actually think our `RefinementRule` is more generic than this, right? It contains a generic grid of the original polytope, which can be cut in whichever way you desire, and contains all the information you have described. 

The refinement described in this particular section (4.1.1) is the one I use in `refine(model::UnstructuredDiscreteModel)`. It is very handy because it allows you to know the ordering of all the new nodes instantly due to clever numbering using not only the coarse node numbering but also the coarse high-dimension face labeling as well. In fact, I have been thinking that this function should stay as it is since, although it is not general for every possible `RefinementRule`, it is quite efficient while remaining very simple.

Reply:
> Another concern that I have here in regards to the refinement of meshes with simplices (triangles, tetrahedra).
In Gridap, the FE machinery, assumes (requires) that simplices meshes are properly oriented. This is achieved by sorting the global vertex identifiers within each cell in increasing order.
See, e.g., https://github.com/gridap/GridapGmsh.jl/blob/master/src/GmshDiscreteModels.jl#L311
Are we fulfilling this condition for the meshes resulting from refinement? (I did not check it)
See also section 4.3 in the following manuscript https://www.sciencedirect.com/science/article/pii/S096599781831113X

This is definitely not achieved here, since as I said in the previous comment the new vertices are ordered globally in the following way: First the old vertices (1:nNodesOld), then vertices created from each edge (nNodesOld+1:nNodesOld+nEdgesOld), then vertices created from each cell (only in case of QUADs). In 3D, we would also have a new vertex for each face. This scheme allows to very easily number the new vertexes in each cell. 

However I don't see a problem in implementing a permutation of the global ids such that we sort ids within each cell. I assume such a sorting always exists, right? Also, this is not needed for mixed (QUAD+TRI) meshes, right?

Reply:
>I actually think our RefinementRule is more generic than this, right? It contains a generic grid of the original polytope, which can be cut in whichever way you desire, and contains all the information you have described.

Yes, I agree. `RefinementRule` seems to be general enough. They only part missing is the fact that we are not leveraging it in the actual refinement process. Another concern is that `RefinementRule` being a unique type, we cannot do type dispatching to implement different strategies, etc. I agree we can leave it as it is by now.

Reply:
> However I don't see a problem in implementing a permutation of the global ids such that we sort ids within each cell. I assume such a sorting always exists, right? Also, this is not needed for mixed (QUAD+TRI) meshes, right?

Let me note this is NOT a global re-labeling of the vertices (just in case there is any confusion, perhaps there is not). It actually is a modification (permutation) of the correspondence among local (to-cell) vertex identifiers and global ID vertex identifiers done cell-wise, cell-by-cell, by sorting global vertex IDs in increasing order.

> I assume such a sorting always exists, right? 

Yes. See code link above. 

>  Also, this is not needed for mixed (QUAD+TRI) meshes, right?

I haven't though much about how this is handled with QUAD+TRIs, not sure if Gridap can handle this either ...